### PR TITLE
[feat] Added VQACPv2 Dataset Implementation

### PR DESCRIPTION
**Pull request by parthsuresh**
Adds support for VQACPv2 dataset in MMF. Automatic downloads are only supported at image level.

Test Plan:
A config has been added to test on MMBT with Images.

`mmf_run config=projects/mmbt/configs/vqacp_v2/with_images.yaml run_type=train_val dataset=vqacp_v2 model=mmbt  training.batch_size=64 training.seed=1`

### CODE REVIEW

File: mmf/datasets/builders/vqacp_v2/builder.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+from mmf.common.registry import registry
+from mmf.datasets.builders.vqacp_v2.dataset import VQACPv2Dataset
+from mmf.datasets.mmf_dataset_builder import MMFDatasetBuilder
+
+
+@registry.register_builder("vqacp_v2")
+class OKVQABuilder(MMFDatasetBuilder):
```
Comment (apsdehal) [inline mmf/datasets/builders/vqacp_v2/builder.py:8]:
```suggestion
class VQACPV2Builder(MMFDatasetBuilder):
```

File: mmf/datasets/builders/vqacp_v2/database.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+from mmf.datasets.databases.annotation_database import AnnotationDatabase
+from mmf.datasets.databases.image_database import (
+    ImageDatabase,
+    get_possible_image_paths,
+)
+from mmf.utils.file_io import PathManager
+
+
+class VQACPv2AnnotationDatabase(AnnotationDatabase):
```
Comment (apsdehal) [inline mmf/datasets/builders/vqacp_v2/database.py:14]: Inherit the OKVQAAnnotationDatabase to reduce duplication.

File: mmf/datasets/builders/vqacp_v2/database.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    def from_path(self, image_path, use_transforms=True):
+        if "train" in self.base_path[0]:
+            train_path = self.base_path[0]
+            val_path = self.base_path[1]
+        else:
+            train_path = self.base_path[1]
+            val_path = self.base_path[0]
+
+        if "train" in image_path:
```
Comment (apsdehal) [inline mmf/datasets/builders/vqacp_v2/database.py:67]: It might be easier to adjust this in the dataset itself in function `preprocess_sample_info`. Then you won't need to update this.

File: mmf/datasets/builders/vqacp_v2/dataset.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+from mmf.common.sample import Sample
+from mmf.common.typings import MMFDatasetConfigType
+from mmf.datasets.builders.vqacp_v2.database import (
+    VQACPv2AnnotationDatabase,
+    VQACPv2ImageDatabase,
+)
+from mmf.datasets.mmf_dataset import MMFDataset
+
+
+class VQACPv2Dataset(MMFDataset):
```
Comment (apsdehal) [inline mmf/datasets/builders/vqacp_v2/dataset.py:15]: If the code is mostly same with respect to OKVQADataset, some refactoring or inheriting would reduce further duplication.

File: mmf/datasets/builders/vqacp_v2/builder.py (hunk 1 above)
Comment (parthsuresh) [reply mmf/datasets/builders/vqacp_v2/builder.py:8]: Sorry, my bad. Fixed.

File: mmf/datasets/builders/vqacp_v2/database.py (hunk 2 above)
Comment (parthsuresh) [reply mmf/datasets/builders/vqacp_v2/database.py:14]: Done, but I still have to override the ` load_annotation_db` function since the format of the annotations is slightly different. So the code size hasn't reduced by much.

> The format of the questions is same as that of the VQA v1 dataset except that this file only contains the "questions" field

File: mmf/datasets/builders/vqacp_v2/database.py (hunk 3 above)
Comment (parthsuresh) [reply mmf/datasets/builders/vqacp_v2/database.py:67]: Thanks for pointing this out, I adjusted it in the `get_image_path` function in the dataset and then removed the VQACPv2ImageDatabase entirely since it was no longer necessary.

File: mmf/datasets/builders/vqacp_v2/dataset.py (hunk 4 above)
Comment (parthsuresh) [reply mmf/datasets/builders/vqacp_v2/dataset.py:15]: It is mostly the same, except for the get_image_path function. I'm not sure how to inherit/refactor it without making some changes to the OKVQADataset. Do you have any suggestions for this?

File: mmf/common/dataset_loader.py [hunk 5]
```diff
@@ -13,7 +13,6 @@ def load_datasets(self):
         self.train_dataset = MultiDatasetLoader("train")
         self.val_dataset = MultiDatasetLoader("val")
         self.test_dataset = MultiDatasetLoader("test")
-
```
Comment (vedanuj) [inline mmf/common/dataset_loader.py:16]: Remove changes from this file.

File: mmf/datasets/builders/vqacp_v2/dataset.py (hunk 4 above)
Comment (vedanuj) [reply mmf/datasets/builders/vqacp_v2/dataset.py:13]: Instead of extending from the `MMFDataset` extend from the `OKVQADataset` and override the necessary functions.

File: mmf/datasets/mmf_dataset.py [hunk 6]
```diff
@@ -63,7 +63,6 @@ def _get_path_based_on_index(self, config, attribute, index):
             raise ValueError(f"No {attribute} present for type {self.dataset_type}")
 
         paths = config[self.dataset_type]
-
```
Comment (vedanuj) [inline mmf/datasets/mmf_dataset.py:66]: Remove

**Comment by parthsuresh** [comment]
Hi, would you be okay with me opening a new PR for this? My git skills are a bit rusty and I seem to have messed my repo up pretty badly.

**Comment by apsdehal** [comment]
Do this:
```
git add remote fr git@github.com:facebookresearch/mmf.git
git fetch fr master
git reset --hard fr/master
git cherry-pick d50a9cf
git add remote local git@github.com:parthsuresh/mmf.git
git push local -f
```

Let me know if this works.

**Comment by parthsuresh** [comment]
I think it's good now? Thank you for your help!

**Comment by parthsuresh** [comment]
Is there anything else needed?

**Comment by apsdehal** [comment]
Hi @parthsuresh,
Not at the moment. We will soon review it again and get back to you.

File: mmf/datasets/builders/vqacp_v2/dataset.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        # coco_split indicates whether the image is from the train or val split of COCO
+        if "train" in coco_split:
+            image_path = f"COCO_train2014_{str(image_id).zfill(12)}.jpg"
+            image_path = os.path.join(train_path, image_path)
+        else:
+            image_path = f"COCO_val2014_{str(image_id).zfill(12)}.jpg"
+            image_path = os.path.join(val_path, image_path)
+        return image_path
+
+    def __getitem__(self, idx: int) -> Type[Sample]:
```
Comment (apsdehal) [inline mmf/datasets/builders/vqacp_v2/dataset.py:49]: Isn't get item mostly same as OKVQA dataset?

Comment (parthsuresh) [reply mmf/datasets/builders/vqacp_v2/dataset.py:49]: `image_path = self.get_image_path(
            sample_info["image_id"], sample_info["coco_split"]
        )
`
`current_sample.image = self.image_db.from_path(image_path)["images"][0]`

It is mostly the same, just these two lines are different.

**Comment by mertxe78** [comment]
> on Phabricator
